plugin.go

```diff
@@ -35,13 +37,13 @@ func init() {
 
 // InputPlugin interface to represent an input fluent-bit plugin.
 type InputPlugin interface {
-	Init(ctx context.Context, conf ConfigLoader) error
+	Init(ctx context.Context, conf ConfigLoader, cmt *cmetrics.Context) error
```

Comment:
Is this somethig the plugin author should use?
If we want to have metrics about a plugin failing or its latency, we can do it internally in the plugin package itself.

And in the case we want to expose these internals, they should be an interface.

But I strongly suggest not exposing it.

Reply:
> Is this somethig the plugin author should use? If we want to have metrics about a plugin failing or its latency, we can do it internally in the plugin package itself.
> 
> And in the case we want to expose these internals, they should be an interface.
> 
> But I strongly suggest not exposing it.

Yes, the plugin author should explicitly be able declare arbitrary metrics, and should be extensible enough. What's your idea around this?

Reply:
@nicolasparada are you thinking to compose a MetricsHandler interface?

Reply:
But yeah, if we want to expose the functionallity it should be through an interface.

> Accept interfaces, return structs.

We might have something like...
```go
type Counter interface {
  Add(delta float64)
}

type Gauge interface {
  Add(delta float64)
  Set(value float64)
}

type Histogram interface {
  Observe(value float64)
}

type Metrics interface {
  NewCounter(name string) Counter
  NewGauge(name string) Gauge
  NewHistogram(name string) Histogram
}
```

So this becomes quite easy for the plugin author.

```go
type myPlugin struct {
  myCounter plugin.Counter
}

func (plug *thePlugin) Init(ctx context.Context, conf plugin.ConfigLoader, metrics plugin.Metrics) error {
	plug.myCounter = metrics.NewCounter("my_counter")
	return nil
}

func (plug thePlugin) Collect(ctx context.Context, ch chan<- plugin.Message) error {
}

Reply:
Yes, @nicolasparada can you implement this? , I will dismiss #16 though, as we will follow the default metrics using the output plugins approach to be consistent.

Reply:
Yeah, I think I can. SHall I do it on top of this PR?

Reply:
Sure @nicolasparada